File: gui/DesignerV1_55/ShowIconDemo/ShowIconDemo.pas
program showicondemo;

Uses exec,intuition,gadtools,graphics,amiga,diskfont,
     workbench,utility,showicondemowin,dos;

Const
  Spaces : string = '                                        ';

Var
 done     : boolean;
 class    : long;
 code     : word;
 pgsel    : pGadget;
 imsg     : pintuimessage;
 dummy    : long;
 awport   : pmsgport;
 id       : long;
 msg      : pappmessage;
 argptr   : pwbarg;
 dirname  : string;
 filename : string;
 argtype  : long;

Procedure ProcessWindowWin0( Class : long ; Code : word ; IAddress : pbyte);
Var
  pgsel : pgadget;
Begin
  Case Class of
    IDCMP_GADGETUP :
      Begin
        { Gadget message, gadget = pgsel. }
        pgsel:=pgadget(iaddress);
      end;
    IDCMP_CLOSEWINDOW :
      Begin
        done:=true;
        { CloseWindow Now }
      end;
    IDCMP_REFRESHWINDOW :
      Begin
        GT_BeginRefresh( Win0);
        { Refresh window. }
        GT_EndRefresh( Win0, True);
      end;
   end;
end;

Begin
  id:=1;
  done:=false;
  if OpenLibs then
	Begin
	  awport:=createmsgport;
	  if awport<>nil then
	    begin
	  	  if OpenWindowWin0(awport,id) then
		    begin
		  	  while (not done) do
				begin
				  dummy:=Wait(bitmask(Win0^.UserPort^.mp_SigBit) or
				              bitmask(awport^.mp_sigbit));
				
				  { Check Window Message Port }
				  
				  imsg:=GT_GetIMsg(Win0^.UserPort);
				  while (imsg <>nil ) do
					begin
					  class:=imsg^.Class;
					  code:=imsg^.Code;
					  pgsel:=pgadget(imsg^.IAddress); { Only reference if it is a gadget message }
					  GT_ReplyIMsg(imsg);
					  ProcessWindowWin0(class, code, pbyte(pgsel));
					  imsg:=GT_GetIMsg(Win0^.UserPort);
					end;
				  
				  { Check AppWindow Port }
				  
				  msg:=pappmessage(GetMsg(awport));
				  while (msg <>nil ) do
					begin
					  
					  { This only handles the first icon in list }
					  { argptr is actually an array of pwbarg }
					  
					  argptr:=msg^.am_arglist;
					    
					  { Get directory lock }
					    
					  dirname:=fexpandlock(argptr^.wa_lock);
					  
					  { Get filename }
					  
					  ctopas(argptr^.wa_name^,filename);
					  
					  ReplyMsg(pmessage(msg));
					  
					  writeln('Dir = '+dirname+copy(spaces,1,40-length(dirname))+'  Name = '+filename);
					  
					  msg:=pappmessage(GetMsg(awport));
					end;
				  
				end;
			  CloseWindowWin0;
		    end
		   else
		    writeln('Cannot open window.');
		  
		  { Make sure message port empty before freeing }
		  
		  msg:=pappmessage(GetMsg(awport));
		  while (msg <>nil ) do
			begin
			  ReplyMsg(pmessage(msg));
			  msg:=pappmessage(GetMsg(awport));
			end;
		  
		  DeleteMsgPort(awport);
		end
	   else
		writeln('Cannot create message port.');
	  CloseLibs;
	end
   else
	writeln('Cannot open libraries.');
end.

File: gui/DesignerV1_55/ShowIconDemo/ShowIconDemo.c
/* Compile me to get full executable. */

#include <stdio.h>
#include <workbench/startup.h>
#include "showicondemowin.c"

int main(void)
{
int done=0;
ULONG class;
UWORD code;
ULONG appid=1;
struct Gadget *pgsel;
struct IntuiMessage *imsg;
struct AppMessage *appmsg;
struct MsgPort *appport;
UBYTE  Buffer[250];

if (OpenLibs()==0)
	{
	if (appport=CreateMsgPort())
		{
		if (OpenWindowWin0(appport,appid)==0)
			{
			while(done==0)
				{
				
				Wait( (1L << Win0->UserPort->mp_SigBit) | (1L << appport->mp_SigBit) );
				
				imsg=GT_GetIMsg(Win0->UserPort);
				while (imsg != NULL )
					{
					class=imsg->Class;
					code=imsg->Code;
					pgsel=(struct Gadget *)imsg->IAddress; /* Only reference if it is a gadget message */
					GT_ReplyIMsg(imsg);
					if (class==IDCMP_CLOSEWINDOW)
						done=1;
					if (class==IDCMP_REFRESHWINDOW)
						{
						GT_BeginRefresh(Win0);
						GT_EndRefresh(Win0, TRUE);
						}
					imsg=GT_GetIMsg(Win0->UserPort);
					}
				
				/* Process appwindow messages */
				
				while (appmsg=(struct AppMessage *)GetMsg(appport))
		            {
		            
		            printf("File = %s\n",appmsg->am_ArgList->wa_Name);
		            
		            NameFromLock( appmsg->am_ArgList->wa_Lock , Buffer , 250);
		            
		            printf("Dir  = %s\n", Buffer);
		            
		            ReplyMsg( (struct Message *) appmsg);
		            }
				
				}
			
			CloseWindowWin0();
			}
		else
			printf("Cannot open window.\n");
	
		/* Clear message port before closing */
		
		while (appmsg=(struct AppMessage *)GetMsg(appport))
		    ReplyMsg( (struct Message *) appmsg);
		
		DeleteMsgPort(appport);
		}
	else
		printf("Cannot create msg port.\n");
	CloseLibs();
	}
else
	printf("Cannot open libraries.\n");
}
